crates/driver/GUIDELINES.md

```diff
@@ -0,0 +1,53 @@
+# Guidelines
+
+High-level guidelines for writing code in a domain-driven style.
+
+## Tests
+
+Prefer to test at the _highest level of abstraction_. This means that the best tests are those
+that interact with the code at the public API layer. E.g., if the code exposes a REST API,
+the tests should be hitting the REST API. Mocking should be brought to a minimum. The
+only components that should be mocked are those components which are _external_ to our system.
+Example: quoting or settlement scoring logic are not external to the system. However, a remote
+solver or an external API provider are external to our system.
+
+There are a few reasons why this is a good approach:
+- The only way to know that the boundary integrations work (e.g. the boundary between your API
+layer and some other layer) is to include that boundary in the tests.
+- Code needs to be refactorable. When tests touch implementation details, those implementation
+details become harder to change. Furthermore, because tests need to change along with the
+implementation details, tests are now useless because you may have made a mistake while changing
+them. Developers are less inclined to refactor and code starts to rot.
+
+It's OK to test implementation details if it's an edge case that's impossible to test from the
+highest layer, or if it's so difficult to test from the highest layer that testing it that way
+becomes impractical.
```

Comment:
Do you think cases like checking algorithms are a correct use of unit tests? We have a path-finding algorithm in the baseline solver, as well as basic algorithms around token input/output calculations for AMMs (which we want to check matches the on-chain version).

In general those tests ☝️ are limited enough in scope that they shouldn't require mocking (and if it does, then it could be a code-smell IMO) and should not be an issue with refactoring as:
- If they are removed, than the tests are removed
- If their interface changes in some fundamental way, then we likely need to generate new test data for these cases anyway

Reply:
IMO local unit tests are fine and I want to keep using them. You can still local unit test an algorithm through its public api. The problem with the mocks shows up when you're using a mock of an interface of a different module and then when someone wants to change that module they have to fix all your tests.